src/common/eeprom.cpp

```diff
@@ -348,35 +346,40 @@ eeprom_init_status_t eeprom_init(void) {
     static eeprom_init_status_t status = EEPROM_INIT_Undefined;
     if (status != EEPROM_INIT_Undefined)
         return status; //already initialized
-    uint16_t version;
-    uint16_t features;
     status = EEPROM_INIT_Normal;
     osSemaphoreDef(eepromSema);
     eeprom_sema = osSemaphoreCreate(osSemaphore(eepromSema), 1);
     st25dv64k_init();
 
-    version = variant8_get_ui16(eeprom_get_var(EEVAR_VERSION));
-    features = (version >= 4) ? variant8_get_ui16(eeprom_get_var(EEVAR_FEATURES)) : 0;
-    if ((version >= EEPROM_FIRST_VERSION_CRC) && !eeprom_check_crc32())
+    //read entire eeprom at once
+    eeprom_vars_t &eevars = eeprom_startup_vars(); //initialize internal static variable, first call must be done now - to ensure interrupt is enabled
+
+    const uint16_t version = eevars.VERSION;
+    const uint16_t features = (version >= 4) ? eevars.FEATURES : 0;
+    if ((version >= EEPROM_FIRST_VERSION_CRC) && !eeprom_check_crc32(eevars))
         status = EEPROM_INIT_Defaults;
     else if ((version != EEPROM_VERSION) || (features != EEPROM_FEATURES)) {
-        if (eeprom_convert_from(version, features) == 0) {
+        if (eeprom_convert_from(eevars) == 0) {
             status = EEPROM_INIT_Defaults;
         } else {
             status = EEPROM_INIT_Upgraded;
         }
     }
-    if (status == EEPROM_INIT_Defaults)
+    switch (status) {
+    case EEPROM_INIT_Defaults:
         eeprom_defaults();
-    //eeprom_print_vars(); this is not possible here because it hangs - init is now done in main.cpp, not in defaultThread
-    eeprom_startup_vars(); //initialize internal static variable, first call must be done now - to ensure interrupt is enabled
+        break;
+    case EEPROM_INIT_Upgraded:
+        eeprom_write_vars(eevars);
+        break;
+    default:
+        break;
+    }
+
     return status;
 }
 
-void eeprom_defaults(void) {
-    eeprom_vars_t vars = eeprom_var_defaults;
-    vars.FWBUILD = project_build_number;
-    vars.FWVERSION = eeprom_fwversion_ui16();
+static void eeprom_write_vars(eeprom_vars_t &vars) {
     eeprom_lock();
     // calculate crc32
     vars.CRC32 = crc32_eeprom((uint32_t *)(&vars), (EEPROM_DATASIZE - 4) / 4);
```

Comment:
This probably requires EEPROM_DATASIZE % 4 == 0, but there is no static_assert anywhere.

Reply:
there is padding variable making so .. not my code
sadly it is different than code in private